salalib/axialmap.cpp

```diff
@@ -3695,9 +3701,11 @@ int ShapeGraph::analyseTulip(Communicator *comm, int tulip_bins, bool choice, in
                         break;
                      }
                      if ((coverage >> rbin) != 0) {
-                        bins[(currentbin + tulip_bins + extradepth) % tulip_bins].add(
-                           SegmentData(conn,SegmentRef(1,lineindex.ref),lineindex.segdepth+1,lineindex.metricdepth+seglength,(coverage >> rbin) << rbin), paftl::ADD_DUPLICATE);
-                        opencount++;
+                         SegmentData sd(conn,SegmentRef(1,lineindex.ref),lineindex.segdepth+1,lineindex.metricdepth+seglength,(coverage >> rbin) << rbin);
+                         size_t bin = (currentbin + tulip_bins + extradepth) % tulip_bins;
+                         auto it = std::lower_bound(bins[bin].begin(), bins[bin].end(), sd);

```

Comment:
You are writing the same non-trivial search and insert here again - this should go into a helper function like `insert_sorted`. See e.g. here for a fairly elegant version: https://stackoverflow.com/questions/15843525/how-do-you-insert-the-value-in-a-sorted-vector

Reply:
Could those bin thingies be sets or multisets?

Reply:
Can not be sets because the original implementation [allowed duplicates](https://github.com/SpaceGroupUCL/depthmapX/pull/215/files#diff-8eeec3bc2e72858dbe7555a41d8ec1b5L3699), but I'll add the helper function